lib/customisation.nix

```diff
@@ -306,18 +306,117 @@ rec {
     in if drv == null then null else
       deepSeq drv' drv';
 
-  /* Make a set of packages with a common scope. All packages called
-     with the provided `callPackage` will be evaluated with the same
-     arguments. Any package in the set may depend on any other. The
-     `overrideScope'` function allows subsequent modification of the package
-     set in a consistent way, i.e. all packages in the set will be
-     called with the overridden packages. The package sets may be
-     hierarchical: the packages in the set are called with the scope
-     provided by `newScope` and the set provides a `newScope` attribute
-     which can form the parent scope for later package sets.
+  /**
+    Make an attribute set (a "scope") from functions that take arguments from that same attribute set.
```

Comment:
I think you did a great job documenting "what" `makeScope` does (it's hard to follow the explanations sometimes due to the high amount of names/concepts involved and some back and forth is needed, but I think that's more due to the complexity of the problem).

I haven't read too much code that uses `makeScope` yet, but I think this documentation is missing a "why would I use this". For example, which use cases/problems does this help me with?

This is more up to you: if you think the audience that comes here will already know _why_ they need something like this, feel free to ignore this. Maybe some other place in the docs or some other different resource would be a better medium to explain why someone would need this. But if that's not the case, I think just mentioning the problems or use cases that this helps with would be already good content.

Reply:
TBH I don't exactly know what it's even for, and it seems like it's just a convenience wrapper around `callPackageWith` to be able to extend those base attribute sets more easily. The only reason I'm documenting this is because I care about `callPackage`. `callPackageWith` is the actual workhorse that sets up the `override` attribute and deals with the mess from `mkDerivation`. And even that is not the full story, because somewhere in `stdenv` we define the cross-compilation splicing.

So before anyone writes something stupid, I'm in favor of just showing *what* it does, because I can barely imagine anyone using it for non-magical reasons or without knowing what exactly they're doing.